---
title: Les rÃ¨gles des Hooks
---

<Intro>

Les Hooks sont dÃ©finis au moyen de fonctions JavaScript, mais reprÃ©sentent un type particulier de logique d'UI rÃ©utilisables, avec des restrictions sur leurs emplacements possibles d'appel.

</Intro>

<InlineToc />

---

## N'appelez des Hooks qu'au niveau racine {/*only-call-hooks-at-the-top-level*/}

Les fonctions dont le nom commence par `use` sont appelÃ©es [*Hooks*](/reference/react) en React.

**N'appelez pas de Hooks au sein de boucles, conditions, fonctions imbriquÃ©es ou blocks `try`/`catch`/`finally` .** Utilisez plutÃ´t les Hooks systÃ©matiquement au niveau racine de votre fonction React, avant tout retour anticipÃ© Ã©ventuel *(early return, NdT)*. Vous ne pouvez appelez des Hooks que pendant que React fait le rendu d'une fonction composantÂ :

* âœ… Appelez-les Ã  la racine du corps d'une [fonction composant](/learn/your-first-component).
* âœ… Appelez-les Ã  la racine du corps d'un [Hook personnalisÃ©](/learn/reusing-logic-with-custom-hooks).

```js{2-3,8-9}
function Counter() {
  // âœ… CorrectÂ : niveau racine dâ€™une fonction composant
  const [count, setCount] = useState(0);
  // ...
}

function useWindowWidth() {
  // âœ… CorrectÂ : niveau racine dâ€™un Hook personnalisÃ©
  const [width, setWidth] = useState(window.innerWidth);
  // ...
}
```

React **ne permet pas** l'appels de Hooks (les fonctions dont le nom commence par `use`) dans d'autres situations, par exempleÂ :

* ğŸ”´ N'appelez pas de Hooks au sein de conditions ou de boucles.
* ğŸ”´ N'appelez pas de Hooks aprÃ¨s une instruction `return` conditionnelle.
* ğŸ”´ N'appelez pas de Hooks au sein de gestionnaires d'Ã©vÃ©nements.
* ğŸ”´ N'appelez pas de Hooks dans des composants Ã  base de classes.
* ğŸ”´ N'appelez pas de Hooks au sein de fonctions passÃ©es Ã  `useMemo`, `useReducer` ou `useEffect`.
* ğŸ”´ N'appelez pas de Hooks au sein de blocs `try`/`catch`/`finally`.

Si vous enfreignez ces rÃ¨gles, vous verrez peut-Ãªtre une erreur.

```js{3-4,11-12,20-21}
function Bad({ cond }) {
  if (cond) {
    // ğŸ”´ IncorrectÂ : au sein dâ€™une condition (pour corriger Ã§a, sortez-en-leÂ !)
    const theme = useContext(ThemeContext);
  }
  // ...
}

function Bad() {
  for (let i = 0; i < 10; i++) {
    // ğŸ”´ IncorrectÂ : au sein dâ€™une boucle (pour corriger Ã§a, sortez-en-leÂ !)
    const theme = useContext(ThemeContext);
  }
  // ...
}

function Bad({ cond }) {
  if (cond) {
    return;
  }
  // ğŸ”´ IncorrectÂ : aprÃ¨s un retour anticipÃ© conditionnel (dÃ©placez-le avant le returnÂ !)
  const theme = useContext(ThemeContext);
  // ...
}

function Bad() {
  function handleClick() {
    // ğŸ”´ IncorrectÂ : au sein dâ€™un gestionnaire dâ€™Ã©vÃ©nements (pour corriger Ã§a, sortez-en-leÂ !)
    const theme = useContext(ThemeContext);
  }
  // ...
}

function Bad() {
  const style = useMemo(() => {
    // ğŸ”´ IncorrectÂ : au sein dâ€™un useMemo (pour corriger Ã§a, sortez-en-leÂ !)
    const theme = useContext(ThemeContext);
    return createStyle(theme);
  });
  // ...
}

class Bad extends React.Component {
  render() {
    // ğŸ”´ IncorrectÂ : au sein dâ€™un composant Ã  base de classe (utilisez plutÃ´t une fonction composantÂ !)
    useEffect(() => {})
    // ...
  }
}

function Bad() {
  try {
    // ğŸ”´ IncorrectÂ : au sein dâ€™un bloc try/catch/finally block (pour corriger Ã§a, sortez-en-leÂ !)
    const [x, setX] = useState(0);
  } catch {
    const [x, setX] = useState(1);
  }
}
```

Vous pouvez utiliser le [plugin `eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks) pour repÃ©rer ces erreurs.

<Note>

[Les Hooks personnalisÃ©s](/learn/reusing-logic-with-custom-hooks) *peuvent* appeler d'autres Hooks (c'est mÃªme leur raison d'Ãªtre). Ã‡a fonctionne parce que les Hooks personnalisÃ©s sont eux aussi censÃ©s n'Ãªtre appelÃ©s que pendant la phase de rendu d'une fonction composant.

</Note>

---

## N'appelez des Hooks que depuis des fonctions React {/*only-call-hooks-from-react-functions*/}

N'appelez pas de Hooks depuis des fonctions JavaScript classiques.  Vous devriez plutÃ´tÂ :

âœ… Appeler des Hooks depuis des fonctions composants React.
âœ… Appeler des Hooks depuis des [Hooks personnalisÃ©s](/learn/reusing-logic-with-custom-hooks#extracting-your-own-custom-hook-from-a-component).

En suivant cette rÃ¨gle, vous garantissez que toute la logique Ã  Ã©tat d'un composant est clairement visible dans son code source.

```js {2,5}
function FriendList() {
  const [onlineStatus, setOnlineStatus] = useOnlineStatus(); // âœ…
}

function setOnlineStatus() { // âŒ Ni un composant ni un Hook personnalisÃ©Â !
  const [onlineStatus, setOnlineStatus] = useOnlineStatus();
}
```
